File: stock_search/models/convex/store.py
"""Persist portfolio, stock, and metadata records through Convex."""

from __future__ import annotations

from typing import Any

from .client import ConvexAPIError, ConvexHttpAdapter
from .convex_schemas import (
    normalize_portfolio_positions,
    normalize_stock_map,
    payload_to_stock_map,
    stock_map_to_rows,
)
from .function_names import (
    CONVEX_META_GET,
    CONVEX_META_SET,
    CONVEX_NEWS_LIST,
    CONVEX_NEWS_REPLACE_ALL,
    CONVEX_PORTFOLIO_GET,
    CONVEX_PORTFOLIO_SET,
    CONVEX_STOCK_GET,
    CONVEX_STOCK_GET_MANY,
    CONVEX_STOCK_LIST,
    CONVEX_STOCK_REPLACE_ALL,
    CONVEX_STOCK_UPSERT,
    CONVEX_STOCK_UPSERT_MANY,
)


class ConvexStore:
    """Typed storage facade for Convex table operations."""

    def __init__(self, *, base_url: str, deploy_key: str) -> None:
        """Initialize the Convex-backed repository wrapper."""
        self._client = ConvexHttpAdapter(
            base_url=base_url,
            deploy_key=deploy_key,
        )

    def load_stocks(self) -> dict[str, dict[str, Any]]:
        """Load the stock map from Convex."""
        payload = self._client.query(CONVEX_STOCK_LIST)
        return payload_to_stock_map(payload)

    @staticmethod
    def _is_missing_function_error(error: ConvexAPIError) -> bool:
        """Return whether Convex rejected a call because the function is undeployed."""
        return "Could not find function" in str(error)

    @staticmethod
    def _normalize_requested_tickers(tickers: list[str]) -> list[str]:
        """Normalize requested tickers while preserving order."""
        return [normalized_ticker for ticker in tickers if (normalized_ticker := str(ticker).strip().upper())]

    @staticmethod
    def _build_upsert_args(row: dict[str, Any]) -> dict[str, Any] | None:
        """Build one stock upsert payload from a partial row."""
        ticker = str(row.get("ticker") or "").strip()
        if not ticker:
            return None

        args: dict[str, Any] = {"ticker": ticker}
        if isinstance(row.get("indicators"), dict):
            args["indicators"] = row["indicators"]
        if isinstance(row.get("evaluation"), dict):
            args["evaluation"] = row["evaluation"]
        if isinstance(row.get("labels"), list):
            args["labels"] = row["labels"]
        return args

    def _load_stocks_one_by_one(
        self,
        tickers: list[str],
    ) -> dict[str, dict[str, Any]]:
        """Load stock rows with single-ticker queries."""
        rows = [self.load_stock(ticker_symbol) for ticker_symbol in tickers]
        return {ticker_symbol: row for ticker_symbol, row in zip(tickers, rows, strict=False) if isinstance(row, dict)}

    def _upsert_stocks_one_by_one(self, rows: list[dict[str, Any]]) -> None:
        """Upsert stock rows through the single-row mutation."""
        for row in rows:
            args = self._build_upsert_args(row)
            if args is None:
                continue
            self._client.mutation(CONVEX_STOCK_UPSERT, args)

    def load_stock(self, ticker: str) -> dict[str, Any] | None:
        """Load one stock row from Convex."""
        payload = self._client.query(CONVEX_STOCK_GET, {"ticker": ticker})
        if not isinstance(payload, dict):
            return None
        mapped = payload_to_stock_map([payload])
        return mapped.get(ticker.upper())

    def load_stocks_by_tickers(self, tickers: list[str]) -> dict[str, dict[str, Any]]:
        """Load multiple stock rows from Convex."""
        normalized_tickers = self._normalize_requested_tickers(tickers)
        if not normalized_tickers:
            return {}
        try:
            payload = self._client.query(
                CONVEX_STOCK_GET_MANY,
                {"tickers": normalized_tickers},
            )
        except ConvexAPIError as error:
            if not self._is_missing_function_error(error):
                raise
            return self._load_stocks_one_by_one(normalized_tickers)
        return payload_to_stock_map(payload)

    def save_stocks(self, stocks_map: dict[str, dict[str, Any]]) -> None:
        """Save stocks."""
        self._client.mutation(
            CONVEX_STOCK_REPLACE_ALL,
            {"rows": stock_map_to_rows(normalize_stock_map(stocks_map))},
        )

    def upsert_stock(
        self,
        *,
        ticker: str,
        indicators: dict[str, Any] | None = None,
        evaluation: dict[str, Any] | None = None,
        labels: list[str] | None = None,
    ) -> None:
        """Upsert one stock row while preserving omitted families."""
        args: dict[str, Any] = {"ticker": ticker}
        if indicators is not None:
            args["indicators"] = dict(indicators)
        if evaluation is not None:
            args["evaluation"] = dict(evaluation)
        if labels is not None:
            args["labels"] = list(labels)
        self._client.mutation(CONVEX_STOCK_UPSERT, args)

    def upsert_stocks(self, rows: list[dict[str, Any]]) -> None:
        """Upsert multiple stock rows while preserving omitted families."""
        normalized_rows = [dict(row) for row in rows if isinstance(row, dict)]
        if not normalized_rows:
            return
        try:
            self._client.mutation(CONVEX_STOCK_UPSERT_MANY, {"rows": normalized_rows})
        except ConvexAPIError as error:
            if not self._is_missing_function_error(error):
                raise
            self._upsert_stocks_one_by_one(normalized_rows)

    def load_portfolio(self, *, key: str = "default") -> dict[str, Any]:
        """Load the portfolio payload from Convex."""
        payload = self._client.query(
            CONVEX_PORTFOLIO_GET,
            {"key": key},
        )
        if not isinstance(payload, dict):
            return {
                "positions": [],
                "portfolio_stats": None,
            }
        positions = payload.get("positions")
        portfolio_stats = payload.get("portfolioStats")
        return {
            "positions": normalize_portfolio_positions(positions if isinstance(positions, list) else []),
            "portfolio_stats": dict(portfolio_stats) if isinstance(portfolio_stats, dict) else None,
        }

    def save_portfolio(
        self,
        *,
        positions: list[dict[str, Any]],
        portfolio_stats: dict[str, Any] | None = None,
        key: str = "default",
    ) -> None:
        """Save portfolio."""
        self._client.mutation(
            CONVEX_PORTFOLIO_SET,
            {
                "key": key,
                "positions": normalize_portfolio_positions(positions),
                "portfolioStats": dict(portfolio_stats) if isinstance(portfolio_stats, dict) else None,
            },
        )

    def load_news(self, *, key: str = "default") -> list[dict[str, Any]]:
        """Load the news feed from Convex."""
        payload = self._client.query(CONVEX_NEWS_LIST, {"key": key})
        return [dict(item) for item in payload if isinstance(item, dict)] if isinstance(payload, list) else []

    def save_news(self, rows: list[dict[str, Any]], *, key: str = "default") -> None:
        """Save news."""
        normalized_rows = [dict(row) for row in rows if isinstance(row, dict)]
        self._client.mutation(
            CONVEX_NEWS_REPLACE_ALL,
            {"key": key, "rows": normalized_rows},
        )

    def get_meta_value(self, key: str) -> str | None:
        """Return meta value."""
        payload = self._client.query(
            CONVEX_META_GET,
            {"key": key},
        )
        if not isinstance(payload, dict):
            return None
        value = payload.get("value")
        return value if isinstance(value, str) else None

    def set_meta_value(self, *, key: str, value: str) -> None:
        """Set meta value."""
        self._client.mutation(
            CONVEX_META_SET,
            {"key": key, "value": value},
        )

File: src/stock-search/models/convex/store.ts
import type { StockAnalysisSectorSnapshot } from "../../data-sources/stockanalysis/index.js";
import { normalizeSectorSnapshot } from "../../data-sources/stockanalysis/sector-cache.js";
import type {
	BackendStore,
	CachedNewsRow,
	PortfolioRecord,
	PositionRow,
	StockEntry,
} from "../../storage/index.js";
import { normalizeTicker } from "../../utils.js";
import { normalizeStockIndicators } from "../schemas.js";
import { ConvexApiError, ConvexHttpClient } from "./client.js";
import {
	CONVEX_PORTFOLIO_POSITION_FIELDS,
	normalizePortfolioPositions,
} from "./convex-schemas.js";
import {
	CONVEX_META_GET,
	CONVEX_META_SET,
	CONVEX_NEWS_DELETE_BY_TICKERS,
	CONVEX_NEWS_LIST,
	CONVEX_NEWS_REPLACE_ALL,
	CONVEX_PORTFOLIO_GET,
	CONVEX_PORTFOLIO_GET_POSITIONS,
	CONVEX_PORTFOLIO_SET,
	CONVEX_PORTFOLIO_SET_POSITIONS,
	CONVEX_REALTIME_TOPICS,
	CONVEX_SECTORS_GET,
	CONVEX_SECTORS_SET,
	CONVEX_STOCK_DELETE_BY_TICKERS,
	CONVEX_STOCK_GET,
	CONVEX_STOCK_GET_MANY,
	CONVEX_STOCK_LIST,
	CONVEX_STOCK_REPLACE_ALL,
	CONVEX_STOCK_UPSERT,
	CONVEX_STOCK_UPSERT_MANY,
} from "./function-names.js";

function normalizeLabels(value: unknown): string[] {
	if (!Array.isArray(value)) {
		return [];
	}
	return value.map((label) => String(label ?? "").trim()).filter(Boolean);
}

function normalizeObject(value: unknown): Record<string, unknown> {
	if (typeof value !== "object" || value === null || Array.isArray(value)) {
		return {};
	}
	return value as Record<string, unknown>;
}

function trimmedString(value: unknown): string | undefined {
	return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

function extraFields(value: Record<string, unknown>): Record<string, unknown> {
	return Object.fromEntries(
		Object.entries(value)
			.filter(([, fieldValue]) => fieldValue !== undefined)
			.filter(([key]) => !CONVEX_PORTFOLIO_POSITION_FIELDS.has(key)),
	);
}

function normalizePosition(value: unknown): PositionRow | null {
	if (typeof value !== "object" || value === null || Array.isArray(value)) {
		return null;
	}
	const row = value as Record<string, unknown>;
	const ticker = normalizeTicker(row.ticker);
	if (!ticker) {
		return null;
	}
	const position: PositionRow = {
		...normalizeObject(row.extra),
		...extraFields(row),
		ticker,
	};
	const quantity = Number(row.quantity);
	if (Number.isFinite(quantity)) {
		position.quantity = quantity;
	}
	const strategy = trimmedString(row.strategy);
	if (strategy) {
		position.strategy = strategy;
	}
	const positionSource = trimmedString(row.position_source);
	if (positionSource) {
		position.position_source = positionSource;
	}
	const industryLabels = normalizeLabels(row.industry_labels);
	if (industryLabels.length > 0) {
		position.industry_labels = industryLabels;
	}
	return position;
}

function normalizePositionRows(value: unknown): PositionRow[] {
	if (!Array.isArray(value)) {
		return [];
	}
	return value
		.map(normalizePosition)
		.filter((position): position is PositionRow => position !== null);
}

function normalizeStockEntry(value: unknown): StockEntry | null {
	if (typeof value !== "object" || value === null || Array.isArray(value)) {
		return null;
	}
	const row = value as Record<string, unknown>;
	return {
		indicators:
			typeof row.indicators === "object" && row.indicators !== null
				? normalizeStockIndicators(row.indicators)
				: {},
		evaluation:
			typeof row.evaluation === "object" && row.evaluation !== null
				? (row.evaluation as Record<string, unknown>)
				: {},
		labels: normalizeLabels(row.labels),
	};
}

function payloadToStockMap(payload: unknown): Record<string, StockEntry> {
	if (!Array.isArray(payload)) {
		return {};
	}
	const rows: Record<string, StockEntry> = {};
	for (const item of payload) {
		const record =
			typeof item === "object" && item !== null
				? (item as Record<string, unknown>)
				: null;
		const ticker = normalizeTicker(record?.ticker);
		const entry = normalizeStockEntry(record);
		if (!ticker || !entry) {
			continue;
		}
		rows[ticker] = entry;
	}
	return rows;
}

export class ConvexStore implements BackendStore {
	readonly backendName = "convex" as const;

	private readonly client: ConvexHttpClient;

	constructor(baseUrl: string, deployKey: string) {
		this.client = new ConvexHttpClient(baseUrl, deployKey);
	}

	private isMissingFunctionError(error: unknown): boolean {
		return (
			error instanceof ConvexApiError &&
			/Could not find(?: public)? function/.test(String(error.message))
		);
	}

	async loadPortfolio(key = "default"): Promise<PortfolioRecord> {
		const payload = await this.client.query<Record<string, unknown> | null>(
			CONVEX_PORTFOLIO_GET,
			{ key },
		);
		return {
			positions: normalizePositionRows(payload?.positions),
			portfolioStats:
				typeof payload?.portfolioStats === "object" &&
				payload.portfolioStats !== null
					? (payload.portfolioStats as Record<string, unknown>)
					: null,
		};
	}

	async savePortfolio({
		positions,
		portfolioStats,
		key = "default",
	}: PortfolioRecord & { key?: string }): Promise<void> {
		await this.client.mutation(CONVEX_PORTFOLIO_SET, {
			key,
			positions: normalizePortfolioPositions(positions),
			portfolioStats,
		});
	}

	async savePortfolioStats(
		portfolioStats: Record<string, unknown> | null,
		key = "default",
	): Promise<void> {
		const existing = await this.loadPortfolio(key);
		await this.savePortfolio({
			key,
			positions: existing.positions,
			portfolioStats,
		});
	}

	async loadPositions(): Promise<PositionRow[]> {
		try {
			const payload = await this.client.query<unknown[]>(
				CONVEX_PORTFOLIO_GET_POSITIONS,
				{ key: "default" },
			);
			return normalizePositionRows(payload);
		} catch (error) {
			if (!this.isMissingFunctionError(error)) {
				throw error;
			}
		}

		const portfolio = await this.loadPortfolio();
		return portfolio.positions;
	}

	async savePositions(positions: PositionRow[]): Promise<void> {
		try {
			await this.client.mutation(CONVEX_PORTFOLIO_SET_POSITIONS, {
				key: "default",
				positions: normalizePortfolioPositions(positions),
			});
			return;
		} catch (error) {
			if (!this.isMissingFunctionError(error)) {
				throw error;
			}
		}

		const existing = await this.loadPortfolio();
		await this.savePortfolio({
			positions,
			portfolioStats: existing.portfolioStats,
		});
	}

	async loadStocks(): Promise<Record<string, StockEntry>> {
		return payloadToStockMap(await this.client.query(CONVEX_STOCK_LIST));
	}

	async loadStocksByTickers(
		tickers: string[],
	): Promise<Record<string, StockEntry>> {
		return payloadToStockMap(
			await this.client.query(CONVEX_STOCK_GET_MANY, { tickers }),
		);
	}

	async loadStock(ticker: string): Promise<StockEntry | null> {
		const payload = await this.client.query<Record<string, unknown> | null>(
			CONVEX_STOCK_GET,
			{ ticker },
		);
		return normalizeStockEntry(payload);
	}

	async upsertStocks(
		rows: Array<{
			ticker: string;
			indicators?: Record<string, unknown>;
			evaluation?: Record<string, unknown>;
			labels?: string[];
		}>,
	): Promise<void> {
		if (rows.length === 0) {
			return;
		}
		const normalizedRows = rows.map((row) => ({
			...row,
			indicators:
				row.indicators === undefined
					? undefined
					: normalizeStockIndicators(row.indicators),
		}));
		try {
			await this.client.mutation(CONVEX_STOCK_UPSERT_MANY, {
				rows: normalizedRows,
			});
			return;
		} catch (error) {
			if (!this.isMissingFunctionError(error)) {
				throw error;
			}
		}

		await Promise.all(
			normalizedRows.map((row) =>
				this.client.mutation(CONVEX_STOCK_UPSERT, row),
			),
		);
	}

	async deleteStocksByTickers(tickers: string[]): Promise<void> {
		if (tickers.length === 0) {
			return;
		}
		try {
			await this.client.mutation(CONVEX_STOCK_DELETE_BY_TICKERS, { tickers });
			return;
		} catch (error) {
			if (!this.isMissingFunctionError(error)) {
				throw error;
			}
		}

		const rows =
			await this.client.query<Record<string, unknown>[]>(CONVEX_STOCK_LIST);
		const removed = new Set(tickers.map(normalizeTicker).filter(Boolean));
		await this.client.mutation(CONVEX_STOCK_REPLACE_ALL, {
			rows: rows.filter((row) => !removed.has(normalizeTicker(row?.ticker))),
		});
	}

	async loadNews(key = "default"): Promise<CachedNewsRow[]> {
		const payload = await this.client.query<unknown[]>(CONVEX_NEWS_LIST, {
			key,
		});
		if (!Array.isArray(payload)) {
			return [];
		}
		return payload
			.filter((row) => typeof row === "object" && row !== null)
			.map((row) => row as Record<string, unknown>)
			.map((row) => ({
				key: String(row.key ?? key),
				ticker: normalizeTicker(row.ticker),
				row:
					typeof row.row === "object" && row.row !== null
						? (row.row as Record<string, unknown>)
						: {},
				updatedAt: Number(row.updatedAt) || 0,
			}));
	}

	async saveNews(rows: CachedNewsRow[], key = "default"): Promise<void> {
		await this.client.mutation(CONVEX_NEWS_REPLACE_ALL, {
			key,
			rows,
		});
	}

	async deleteNewsByTickers(tickers: string[], key = "default"): Promise<void> {
		if (tickers.length === 0) {
			return;
		}
		try {
			await this.client.mutation(CONVEX_NEWS_DELETE_BY_TICKERS, {
				key,
				tickers,
			});
			return;
		} catch (error) {
			if (!this.isMissingFunctionError(error)) {
				throw error;
			}
		}

		const rows = await this.loadNews(key);
		const removed = new Set(tickers.map(normalizeTicker).filter(Boolean));
		await this.saveNews(
			rows.filter((row) => !removed.has(normalizeTicker(row.ticker))),
			key,
		);
	}

	async loadSectorSnapshot(
		key = "default",
	): Promise<StockAnalysisSectorSnapshot | null> {
		try {
			return normalizeSectorSnapshot(
				await this.client.query(CONVEX_SECTORS_GET, { key }),
			);
		} catch (error) {
			if (this.isMissingFunctionError(error)) {
				return null;
			}
			throw error;
		}
	}

	async saveSectorSnapshot(
		snapshot: StockAnalysisSectorSnapshot,
		key = "default",
	): Promise<void> {
		await this.client.mutation(CONVEX_SECTORS_SET, {
			key,
			sectors: snapshot.sectors,
			meta: snapshot.meta,
		});
	}

	async getMetaValue(key: string): Promise<string | null> {
		const payload = await this.client.query<Record<string, unknown> | null>(
			CONVEX_META_GET,
			{ key },
		);
		return typeof payload?.value === "string" ? payload.value : null;
	}

	async setMetaValue(key: string, value: string): Promise<void> {
		await this.client.mutation(CONVEX_META_SET, { key, value });
	}
}

export const convexRealtimeTopics = [...CONVEX_REALTIME_TOPICS];
